### PR TITLE
Sunset Legacy Sync

### LINKED ISSUE #1596: Fast Sync Must Fetch Deploy Hashes And Save Going Back to The TTL

### LINKED ISSUE #1664: Refactor Chainspec Loader to Restart Daemon
Since #1666, the Chainspec loader does not need to run genesis or upgrades.

The Chainspec loader is to be refactored into a restart daemon.

### LINKED ISSUE #1676: Sunset Legacy Sync

### LINKED ISSUE #1730: Drop `state` Database
Fast sync does not need the `state` database in LMDB, so a data migration to drop that database is necessary.

### PR DESCRIPTION
**Pull request by xcthulhu**
- [x] Closes #1676 
  - [x] Updated `CHANGELOG.md`
  - Other resolved issues
    - [x] Run deploys back to TTL (Closes #1596)
    - [x] Removes `State` database (Closes #1730)

### CODE REVIEW

File: node/src/types/exit_status.rs [hunk 1]
```diff
@@ -1,17 +1,12 @@
...
-const SIGNAL_OFFSET: u8 = 128;
 
 /// Exit codes which should be used by the casper-node binary, and provided by the initializer
 /// reactor to the binary.
 ///
 /// Note that a panic will result in the Rust process producing an exit code of 101.
 #[derive(Clone, Copy, PartialEq, Eq, Debug, DataSize)]
 #[repr(u8)]
-pub enum ExitCode {
+pub enum ExitStatus {
```
Comment (Fraser999) [inline node/src/types/exit_status.rs]: I'm not following why this change was made, and especially why it's part of this PR?

I'd prefer to have this reverted.

File: node/CHANGELOG.md [hunk 2]
```diff
@@ -30,11 +30,15 @@ All notable changes to this project will be documented in this file.  The format
...
 
+### Deprecated
+* The `starting_state_root_hash` field from the REST status endpoint has been deprecated
 
 ### Removed
 * The unofficial support for nix-related derivations and support tooling has been removed.
 * Experimental, nix-based kubernetes testing support has been removed.
 * Removed dead code revealed by making modules `pub(crate)`.
+* Removed undocumented exit codes corresponding to `SIGINT`, `SIGTERM`, etc.
+* Removed legacy `LinearChainSync`
```
Comment (afck) [inline node/CHANGELOG.md]: The user won't know what that means. Maybe:

```suggestion
* Removed legacy synchronization from genesis in favor of fast sync.
```

File: node/src/components/chainspec_loader.rs [hunk 3]
```diff
@@ -81,19 +63,9 @@ pub(crate) enum Event {
...
-            } => {
-                write!(
-                    formatter,
-                    "initialize(maybe_highest_block: {})",
-                    maybe_highest_block
-                        .as_ref()
-                        .map_or_else(|| "None".to_string(), |block| block.to_string())
-                )
+            Event::Initialize => {
+                write!(formatter, "initialize",)
```
Comment (afck) [inline node/src/components/chainspec_loader.rs]:
```suggestion
                write!(formatter, "initialize")
```

(Or even without `{}`, like the next case.)

File: node/src/components/chainspec_loader.rs [hunk 4]
```diff
@@ -246,17 +207,10 @@ impl ChainspecLoader {
             false
         };
 
-        // In case this is a version which should be immediately replaced by the next version, don't
-        // create any effects so we exit cleanly for an upgrade without touching the storage
-        // component.  Otherwise create effects which will allow us to initialize properly.
```
Comment (afck) [inline node/src/components/chainspec_loader.rs:251]: Is that comment really obsolete?

File: node/src/reactor.rs [hunk 5]
```diff
@@ -705,13 +704,13 @@ where
...
-                        break ReactorExit::ProcessShouldExit(ExitCode::Abort);
+                        break ReactorExit::ProcessShouldExit(ExitStatus::Abort);
                     }
                 }
-                SIGINT => break ReactorExit::ProcessShouldExit(ExitCode::SigInt),
-                SIGQUIT => break ReactorExit::ProcessShouldExit(ExitCode::SigQuit),
-                SIGTERM => break ReactorExit::ProcessShouldExit(ExitCode::SigTerm),
-                _ => error!("should be unreachable - bug in signal handler"),
+                signal_code => {
+                    error!("Received signal code {}, shutting done", signal_code);
```
Comment (afck) [inline node/src/reactor.rs]:
```suggestion
                    error!("Received signal code {}, shutting down", signal_code);
```

File: node/src/reactor.rs [hunk 6]
```diff
@@ -705,13 +704,13 @@ where
...
+                    break ReactorExit::ProcessShouldExit(ExitStatus::Abort);
+                }
```
Comment (afck) [inline node/src/reactor.rs]: Isn't it slightly safer to keep the old behavior, where we don't shut down on an undefined signal?

File: node/src/components/storage.rs [hunk 7]
```diff
@@ -130,17 +127,6 @@ const STORAGE_FILES: [&str; 5] = [
     "sse_index",
 ];
 
-#[derive(Debug, From, Serialize)]
-#[repr(u8)]
-pub(crate) enum Event {
```
Comment (afck) [inline node/src/components/storage.rs:135]: Should we replace this with `StorageRequest` then, i.e. move what's currently `StorageRequest` here and rename it to `Event`?

File: node/src/reactor/joiner.rs [hunk 8]
```diff
@@ -509,52 +488,18 @@ impl reactor::Reactor for Reactor {
...
-            "block_header_and_finality_signatures_by_height",
-            config.fetcher,
-            registry,
-        )?;
-
+        let block_fetcher = Fetcher::new("block", config.fetcher, registry)?;
+        let trie_fetcher = Fetcher::new("trie", config.fetcher, registry)?;
+        let block_header_and_finality_signatures_by_height_fetcher =
+            Fetcher::new("block_header_by_height", config.fetcher, registry)?;
         let block_header_by_hash_fetcher: Fetcher<BlockHeader> =
```
Comment (afck) [inline node/src/reactor/joiner.rs:497]: Should this just be `block_header_fetcher` then, since for `block_fetcher` it's also implicit?

File: node/src/reactor/joiner.rs [hunk 9]
```diff
@@ -852,63 +771,8 @@ impl reactor::Reactor for Reactor {
...
-            JoinerEvent::LinearChain(event) => reactor::wrap_effects(
-                JoinerEvent::LinearChain,
-                self.linear_chain.handle_event(effect_builder, rng, event),
-            ),
-            JoinerEvent::BlockProposerRequest(request) => {
-                // Consensus component should not be trying to create new blocks during joining
-                // phase.
-                error!("ignoring block proposer request {}", request);
+            JoinerEvent::ContractRuntimeAnnouncement(_) => {
+                // Not used
```
Comment (afck) [inline node/src/reactor/joiner.rs]:
```suggestion
```

(E.g. `NetworkAnnouncement` also doesn't have a comment.)

File: node/src/components/chainspec_loader.rs [hunk 10]
```diff
@@ -223,11 +187,8 @@ impl ChainspecLoader {
             let chainspec_loader = ChainspecLoader {
                 chainspec,
                 root_dir,
-                reactor_exit: Some(ReactorExit::ProcessShouldExit(ExitCode::Abort)),
```
Comment (afck) [inline node/src/components/chainspec_loader.rs:226]: (Linking #1966 for reference: A lot of the logic removed here will have to be reimplemented in #1966.)

File: node/src/components/linear_chain_sync/operations.rs [hunk 11]
```diff
@@ -484,6 +472,26 @@ where
         }
     }
 
+    // Fetch and store all blocks that can contain not-yet-expired deploys. These are needed for
+    // replay detection.
+    let mut current_header = most_recent_block_header.clone();
+    while current_header.timestamp().elapsed() < chainspec.deploy_config.max_ttl
```
Comment (afck) [inline node/src/components/linear_chain_sync/operations.rs]: (Not your fault — mine, actually!)
I wonder whether this should be

```suggestion
    while era_start.saturating_diff(current_header.timestamp()) < chainspec.deploy_config.max_ttl
```

instead, where `era_start` is the start of the most recent era we know of?

Because while we'll only _propose_ blocks with future timestamps, we might have to validate proposals with a timestamp older than that, but never older than the beginning of the most recent era.

Not sure if we have that era start information readily available here — otherwise it should probably be a separate ticket.

**Review by xcthulhu — COMMENTED** [review]
@goral09 I've added some changes from #1995.  If you prefer I can defer those to #1995 or another PR.

File: node/CHANGELOG.md [hunk 12]
```diff
@@ -30,11 +30,15 @@ All notable changes to this project will be documented in this file.  The format
 * More node modules are now `pub(crate)`.
 * Chain automatically creates a switch block immediately after genesis or an upgrade.
 
+### Deprecated
+* The `starting_state_root_hash` field from the REST status endpoint has been deprecated
```
Comment (Fraser999) [inline node/CHANGELOG.md]:
```suggestion
* Deprecate the `starting_state_root_hash` field from the REST and JSON-RPC status endpoints.
```

File: node/CHANGELOG.md [hunk 13]
```diff
@@ -30,11 +30,15 @@ All notable changes to this project will be documented in this file.  The format
...
 
 ### Removed
 * The unofficial support for nix-related derivations and support tooling has been removed.
 * Experimental, nix-based kubernetes testing support has been removed.
 * Removed dead code revealed by making modules `pub(crate)`.
+* Removed undocumented exit codes corresponding to `SIGINT`, `SIGTERM`, etc.
+* Removed legacy synchronization from genesis in favor of fast sync.
```
Comment (Fraser999) [inline node/CHANGELOG.md]:
```suggestion
* Remove dead code revealed by making modules `pub(crate)`.
* Remove undocumented exit codes corresponding to `SIGINT`, `SIGTERM`, etc.
* Remove legacy synchronization from genesis in favor of fast sync.
```

File: node/CHANGELOG.md [hunk 14]
```diff
@@ -30,11 +30,15 @@ All notable changes to this project will be documented in this file.  The format
...
 * Chain automatically creates a switch block immediately after genesis or an upgrade.
 
+### Deprecated
+* The `starting_state_root_hash` field from the REST status endpoint has been deprecated
 
 ### Removed
 * The unofficial support for nix-related derivations and support tooling has been removed.
 * Experimental, nix-based kubernetes testing support has been removed.
 * Removed dead code revealed by making modules `pub(crate)`.
+* Removed undocumented exit codes corresponding to `SIGINT`, `SIGTERM`, etc.
```
Comment (Fraser999) [inline node/CHANGELOG.md]: I'd prefer if this change was reverted, meaning this bullet point should also be removed.

If we decide to not revert those changes, then let's remove the word "undocumented" as they are currently documented.  Also, let's replace ", etc" with a complete list.

File: node/src/reactor/joiner/memory_metrics.rs [hunk 15]
```diff
@@ -146,11 +146,9 @@ impl MemoryMetrics {
         let chainspec_loader = reactor.chainspec_loader.estimate_heap_size() as i64;
         let storage = reactor.storage.estimate_heap_size() as i64;
         let contract_runtime = reactor.contract_runtime.estimate_heap_size() as i64;
-        let linear_chain_fetcher = reactor.linear_chain_fetcher.estimate_heap_size() as i64;
+        let block_fetcher = reactor.block_fetcher.estimate_heap_size() as i64;
         let linear_chain_sync = reactor.linear_chain_sync.estimate_heap_size() as i64;
-        let block_validator = reactor.block_validator.estimate_heap_size() as i64;
         let deploy_fetcher = reactor.deploy_fetcher.estimate_heap_size() as i64;
-        let linear_chain = reactor.linear_chain.estimate_heap_size() as i64;
```
Comment (Fraser999) [inline node/src/reactor/joiner/memory_metrics.rs]: Shouldn't the `block_validator` and `linear_chain` metrics just be removed from `MemoryMetrics` altogether?

File: node/src/types/status_feed.rs [hunk 16]
```diff
@@ -173,11 +164,12 @@ pub struct GetStatusResult {
 }
 
 impl GetStatusResult {
+    #[allow(deprecated)]
     pub(crate) fn new(status_feed: StatusFeed<NodeId>, api_version: ProtocolVersion) -> Self {
         GetStatusResult {
             api_version,
             chainspec_name: status_feed.chainspec_info.name,
-            starting_state_root_hash: status_feed.chainspec_info.starting_state_root_hash,
+            starting_state_root_hash: Some(Default::default()),
```
Comment (Fraser999) [inline node/src/types/status_feed.rs:172]: I'm still unclear on why we lose the ability to set this value due to fast-sync.  It'd be good to discuss offline to see if there's a way to retain the status quo, or at least have the value set properly once fast-sync has finished syncing.

Also, it's not clear why the type has changed to `Option<Digest>` if we only ever set it to `Some`.

File: node/src/reactor/joiner.rs [hunk 17]
```diff
@@ -1024,13 +872,15 @@ impl reactor::Reactor for Reactor {
...
             ),
+            JoinerEvent::FinishedJoining { block_header } => {
+                self.linear_chain_sync = LinearChainSyncState::Done(block_header);
+                Effects::new()
+            }
         }
     }
 
     fn maybe_exit(&self) -> Option<ReactorExit> {
-        if self.linear_chain_sync.stopped_for_upgrade() {
```
Comment (Fraser999) [inline node/src/reactor/joiner.rs:1031]: Is this not still a possible exit point - i.e. while syncing we hit the next activation point?

File: node/src/components/chainspec_loader.rs (hunk 4 above)
Comment (Fraser999) [reply node/src/components/chainspec_loader.rs:251]: Yes, I think so.  In fact, I think we should remove the `Initialize` variant from the `Event` enum as it seems like it's doing nothing now.

**Review by Fraser999 — COMMENTED** [review]
Looks like the `linear_chain_sync` section should be removed from our various config TOML files.

File: node/CHANGELOG.md (hunk 14 above)
Comment (xcthulhu) [reply node/CHANGELOG.md]: I removed this because as far as I can tell the launcher is not using them. These are also non-standard... YAGNI?

Comment (Fraser999) [reply node/CHANGELOG.md]: The launcher uses `SIGTERM` to kill the node if it itself panics or receives one of the three termination signals.

> These are also non-standard

How so?  They're standard Unix signals.

If this change isn't really required by fast-sync, then it shouldn't be part of this PR.

File: node/src/reactor/joiner.rs (hunk 17 above)
Comment (xcthulhu) [reply node/src/reactor/joiner.rs:1031]: This is no longer possible, since I do not expose this state.

File: node/src/components/chainspec_loader.rs (hunk 4 above)
Comment (xcthulhu) [reply node/src/components/chainspec_loader.rs:251]: Sadly, the `initializer` joiner won't progress without this component sending an `Initialize` event to itself, even if it's a no-op.

File: node/src/reactor.rs (hunk 6 above)
Comment (xcthulhu) [reply node/src/reactor.rs]: Looks like this is a blocker for @Fraser999

File: node/src/components/chainspec_loader.rs (hunk 4 above)
Comment (xcthulhu) [reply node/src/components/chainspec_loader.rs:251]: @Fraser999 we should circle back on this...

File: node/src/reactor/joiner.rs (hunk 8 above)
Comment (xcthulhu) [reply node/src/reactor/joiner.rs:497]: @afck I can try to resolve this

File: node/src/components/linear_chain_sync/operations.rs (hunk 11 above)
Comment (xcthulhu) [reply node/src/components/linear_chain_sync/operations.rs]: Looks like this is a blocker; `era_start` here is the trusted-key-block timestamp.

File: node/src/types/exit_status.rs (hunk 1 above)
Comment (xcthulhu) [reply node/src/types/exit_status.rs]: So right now the restart logic doesn't work for fast-sync... I'll make this revert and I'll ticket how we anticipate fast sync working.

File: node/CHANGELOG.md (hunk 14 above)
Comment (xcthulhu) [reply node/CHANGELOG.md]: Okay, it looks like we are following a *nix convention here, and this exit code is getting logged in the launcher: https://github.com/casper-network/casper-node-launcher/blob/main/src/utils.rs#L117

I am non-plussed about how the launcher exit codes are different, but we can leave this in.

File: node/src/reactor/joiner/memory_metrics.rs (hunk 15 above)
Comment (xcthulhu) [reply node/src/reactor/joiner/memory_metrics.rs]: This is an oversight, will fix.

File: node/src/components/linear_chain_sync/operations.rs (hunk 11 above)
Comment (xcthulhu) [reply node/src/components/linear_chain_sync/operations.rs]: Fixed in fd7afac00bc4bef26ad730b1a74fddd3908e7c27

File: node/src/reactor.rs (hunk 6 above)
Comment (xcthulhu) [reply node/src/reactor.rs]: Fixed in d9597e7b74b7776cb33b6cbaadb9afa8cae1d416

File: node/src/types/exit_status.rs (hunk 1 above)
Comment (xcthulhu) [reply node/src/types/exit_status.rs]: Fixed in d9597e7b74b7776cb33b6cbaadb9afa8cae1d416

File: node/CHANGELOG.md (hunk 14 above)
Comment (xcthulhu) [reply node/CHANGELOG.md]: Fixed in d9597e7b74b7776cb33b6cbaadb9afa8cae1d416

File: node/src/reactor/joiner/memory_metrics.rs (hunk 15 above)
Comment (xcthulhu) [reply node/src/reactor/joiner/memory_metrics.rs]: Fixed in bc0eb439142e1eba54368fbe3cb1d0ce835156b6

File: node/src/types/status_feed.rs (hunk 16 above)
Comment (xcthulhu) [reply node/src/types/status_feed.rs:172]: As discussed, this is being deprecated.

We can change this to `None`

Comment (xcthulhu) [reply node/src/types/status_feed.rs:172]: I've documented that we should remove this in 2.0.0 in https://github.com/casper-network/casper-node/issues/2023

Comment (xcthulhu) [reply node/src/types/status_feed.rs:172]: ~I changed this to `None` in 6fc17a4fd320d8e2cc910e2abc7dcb9dfa3dbc17~

[EDIT: Since this is always `None`, it will never show up in the client and so I've removed it altogether in 3d2f7c1e110fd1a6a41cb07b45cc6e2851156e89]

[EDIT2: Looks like we can't just get rid of it for API compatibility reasons, so I've reverted removing this.]

File: node/CHANGELOG.md [hunk 18]
```diff
@@ -30,11 +30,14 @@ All notable changes to this project will be documented in this file.  The format
 * More node modules are now `pub(crate)`.
 * Chain automatically creates a switch block immediately after genesis or an upgrade.
 
+### Deprecated
+* Deprecate the `starting_state_root_hash` field from the REST and JSON-RPC status endpoints.
```
Comment (goral09) [inline node/CHANGELOG.md:35]: It's now _removed_ I suppose.

File: node/CHANGELOG.md [hunk 19]
```diff
@@ -80,8 +83,6 @@ All notable changes to this project will be documented in this file.  The format
 * Limit the maximum number of clients connected to the event stream server via the `[event_stream_server][max_concurrent_subscribers]` config option.
 * Avoid emitting duplicate events in the event stream.
 * Change `BlockIdentifier` params in the Open-RPC schema to be optional.
-* Asymmetric connections are now swept regularly again.
```
Comment (goral09) [inline node/CHANGELOG.md:83]: What happened here?

File: node/src/components/consensus/error.rs [hunk 20]
```diff
@@ -0,0 +1,38 @@
...
+pub enum FinalitySignatureError {
+    #[error(
+        "Block signatures contain bogus validator. \
+         trusted validator weights: {trusted_validator_weights:?}, \
+         block signatures: {block_signatures:?}, \
+         bogus validator public key: {bogus_validator_public_key:?}"
+    )]
+    BogusValidator {
+        trusted_validator_weights: BTreeMap<PublicKey, U512>,
+        block_signatures: Box<BlockSignatures>,
```
Comment (goral09) [inline node/src/components/consensus/error.rs:19]: Opinion: I don't think we need all this information in the error.

File: node/src/components/consensus/error.rs [hunk 21]
```diff
@@ -0,0 +1,38 @@
...
+         signature weight: {signature_weight}, \
+         total validator weight: {total_validator_weight}, \
+         finality threshold fraction: {finality_threshold_fraction}"
+    )]
+    InsufficientWeightForFinality {
+        trusted_validator_weights: BTreeMap<PublicKey, U512>,
+        block_signatures: Box<BlockSignatures>,
+        signature_weight: Box<U512>,
+        total_validator_weight: Box<U512>,
+        finality_threshold_fraction: Ratio<u64>,
```
Comment (goral09) [inline node/src/components/consensus/error.rs:36]: Opinion: IMO these three fields should suffice.

**Review by goral09 — COMMENTED** [review]
I see that you removed the `ChainspecLoader` entirely. Also, `ContractRuntimeRequest::CommitGenesis` and `ContractRuntimeRequest::Upgrade` events (and their handling logic) have been removed. Can you include the explanation of it in the description? Specifically, asnwer questions like:
* Why we're doing it? 
* Why we _can_ do it? Why it is OK to do that.
* What is it being replaced with?
* etc. anything you think is worth adding for people to understand this change.

File: node/src/components/linear_chain_sync/operations.rs [hunk 22]
```diff
@@ -33,24 +30,23 @@ const TIMEOUT_DURATION: Duration = Duration::from_millis(100);
...
+) -> Result<Box<T>, FetcherError<T, NodeId>>
 where
     T: Item + 'static,
-    REv: From<FetcherRequest<I, T>> + From<NetworkInfoRequest<I>>,
-    I: Eq + Debug + Clone + Send + 'static,
+    JoinerEvent: From<FetcherRequest<NodeId, T>>,
 {
     loop {
-        for peer in effect_builder.get_peers_in_random_order().await {
+        'peer_loop: for peer in effect_builder.get_peers_in_random_order().await {
```
Comment (goral09) [inline node/src/components/linear_chain_sync/operations.rs:42]: This will request the same elements from every peer we are connected to. Was that intentional? If so, I don't think it's a good decision.

File: node/src/components/linear_chain_sync/operations.rs [hunk 23]
```diff
@@ -80,7 +76,7 @@ where
                         "Peer timed out",
                     );
                     // Peer timed out fetching the item, continue with the next peer
-                    continue;
+                    continue 'peer_loop;
```
Comment (goral09) [inline node/src/components/linear_chain_sync/operations.rs]: Why did you put `continue` in the `TimedOut` case explicitly but not in `Absent`?

File: node/src/effect/requests.rs [hunk 24]
```diff
@@ -445,44 +429,6 @@ impl Display for StorageRequest {
...
-    },
-    #[cfg(test)]
-    /// Loads a piece of state from storage.
-    Load {
-        /// Key to load from.
-        key: Cow<'static, [u8]>,
-        /// Responder for value, if found, returning the previously passed in serialization form.
-        responder: Responder<Option<Vec<u8>>>,
-    },
-}
```
Comment (goral09) [inline node/src/effect/requests.rs:470]: I think this code was useful to keep around, even if not used. It's a feature that we may need in the future and I don't want to spend time reinventing/reimplementing it. If anything, removing it shouldn't be done in this PR.

File: node/src/reactor/joiner.rs (hunk 17 above)
Comment (goral09) [reply node/src/reactor/joiner.rs:1031]: Is it true that the new code won't be able to synchronize across upgrades? It will either sync in the era running previous version and then upgrade via the `LinearChain`-raised event or not sync at all if other validators have already moved on and there will be no one to respond to protocol state requests.

File: node/src/effect/requests.rs (hunk 24 above)
Comment (xcthulhu) [reply node/src/effect/requests.rs:470]: Well it's dead code; we can revert it to get it back later if we need it, but for now let's remove it.

File: node/CHANGELOG.md (hunk 18 above)
Comment (xcthulhu) [reply node/CHANGELOG.md:35]: As per discussion, this is apparently a breaking change to remove it so I've reverted.

File: node/src/reactor/joiner.rs (hunk 17 above)
Comment (xcthulhu) [reply node/src/reactor/joiner.rs:1031]: > Is it true that the new code won't be able to synchronize across upgrades?

So right now this isn't implemented but there's a ticket #1966 

In general, you won't always be able to sync across upgrades.  There could be a non-backward-compatible change to the wire format or networking protocol for instance.

File: node/CHANGELOG.md (hunk 19 above)
Comment (xcthulhu) [reply node/CHANGELOG.md:83]: These were fixed but not in 1.3.0 (it's broken there).

Looks like I should move this up to the `Unreleased` section.

Comment (xcthulhu) [reply node/CHANGELOG.md:83]: Fixed in d6c33b569

File: node/src/components/consensus/error.rs (hunk 20 above)
Comment (xcthulhu) [reply node/src/components/consensus/error.rs:19]: It's been helpful to me...

File: node/src/components/consensus/error.rs (hunk 21 above)
Comment (xcthulhu) [reply node/src/components/consensus/error.rs:36]: I have found it helpful to provide this other context when I hit an error.

File: node/src/reactor/joiner.rs (hunk 8 above)
Comment (xcthulhu) [reply node/src/reactor/joiner.rs:497]: I opted to make `block_fetcher` be `block_by_hash_fetcher` instead (since people like me might be confused about how you querying for them exactly).

Fixed in 95710d4a6

File: node/src/components/storage.rs (hunk 7 above)
Comment (xcthulhu) [reply node/src/components/storage.rs:135]: I did something similar to `ContractRuntimeRequest` but I just left it in `effect::requests`

**Comment by xcthulhu** [comment]
> Looks like the `linear_chain_sync` section should be removed from our various config TOML files.

Fixed in f9dbdd698

File: node/src/components/linear_chain_sync/operations.rs (hunk 23 above)
Comment (xcthulhu) [reply node/src/components/linear_chain_sync/operations.rs]: Good catch!  This `continue` doesn't appear to be necessary. In both the `TimeOut` case and the `Absent` case we'll continue with loop after the code moves on from this `match` block.

Comment (xcthulhu) [reply node/src/components/linear_chain_sync/operations.rs]: Fixed in 9bdb3da57

File: node/src/components/consensus/error.rs (hunk 20 above)
Comment (fizyk20) [reply node/src/components/consensus/error.rs:19]: I'm for having it here - if we get that error in production, we better know exactly what signature was the problem, who signed it and whether they truly weren't a validator.

File: node/src/components/consensus/error.rs (hunk 21 above)
Comment (fizyk20) [reply node/src/components/consensus/error.rs:36]: I'm of the same opinion as above: when we actually get the error and want to debug, we'll want as much context as possible.

File: node/src/types/status_feed.rs (hunk 16 above)
Comment (xcthulhu) [reply node/src/types/status_feed.rs:172]: > I see that you removed the `ChainspecLoader` entirely. Also, `ContractRuntimeRequest::CommitGenesis` and `ContractRuntimeRequest::Upgrade` events (and their handling logic) have been removed. Can you include the explanation of it in the description? Specifically, asnwer questions like:
> 
> * Why we're doing it?
> * Why we _can_ do it? Why it is OK to do that.

Nothing is constructing these effects, so it's dead code.

> * What is it being replaced with?

We're calling the genesis and upgrade methods directly in the participating reactor.

This was done in #1666 so we could construct a switch block immediately after genesis/upgade.

#1666 was done because fast-sync's security model won't permit joining a network after an emergency upgrade until it can find a switchblock.  So we'd be in a scenario where we'd have to wait two hours before anybody could get a trusted switch block in that case.

In #1666 @afck and I needed to initialize the `EraSupervisor` properly.  The simplest solution was just to call the contract runtime directly in the participating reactor constructor.

File: node/src/reactor.rs [hunk 25]
```diff
@@ -666,7 +666,6 @@ where
         };
         if let Err(error) = self.scheduler.debug_dump(&mut file).await {
             warn!(%error, "could not serialize debug snapshot to {}", debug_dump_filename);
-            return;
```
Comment (Fraser999) [inline node/src/reactor.rs:669]: Was this change deliberate?

File: node/src/reactor/joiner/memory_metrics.rs (hunk 15 above)
Comment (Fraser999) [reply node/src/reactor/joiner/memory_metrics.rs]: So I think this is still a bit incorrect.  Seems like now:
* we have no `mem_linear_chain_sync` member, but its size is gathered and included in the `total` in the `estimate()` call
* we still have a `mem_linear_chain` member, but it is never updated in the `estimate()` call

File: node/src/components/linear_chain_sync/operations.rs [hunk 26]
```diff
@@ -33,14 +30,13 @@ const TIMEOUT_DURATION: Duration = Duration::from_millis(100);
...
-    effect_builder: EffectBuilder<REv>,
+async fn fetch_retry_forever<T>(
+    effect_builder: EffectBuilder<JoinerEvent>,
     id: T::Id,
-) -> Result<Box<T>, FetcherError<T, I>>
+) -> Result<Box<T>, FetcherError<T, NodeId>>
 where
     T: Item + 'static,
-    REv: From<FetcherRequest<I, T>> + From<NetworkInfoRequest<I>>,
-    I: Eq + Debug + Clone + Send + 'static,
```
Comment (fizyk20) [inline node/src/components/linear_chain_sync/operations.rs:43]: Is it a good idea to lose the generic ID type? It tends to be useful in testing. Unless we're sure that we won't be adding tests of these methods with some simpler node IDs?

Similar for `REv` in other methods below.

File: utils/nctl/sh/scenarios/configs/gov96.config.toml [hunk 27]
```diff
@@ -316,8 +316,4 @@ verify_accounts = true
 # Optional depth limit to use for global state queries.
 #
 # If unset, defaults to 5.
-#max_query_depth = 5
-
-[linear_chain_sync]
-# Use sync_timeout to set the amount of time that the node will try to sync before shutting down.
-sync_timeout = "1hr"
+#max_query_depth = 5
```
Comment (Fraser999) [inline utils/nctl/sh/scenarios/configs/gov96.config.toml]: I think @mpapierski discovered recently that this file needs to remain compatible with whatever version our mainnet nodes are running, so this change will likely need to be reverted.

File: node/src/effect/requests.rs (hunk 24 above)
Comment (Fraser999) [reply node/src/effect/requests.rs:470]: I prefer to remove code too and rely on VCS history rather than leave it in the codebase.

File: node/src/components/chainspec_loader.rs (hunk 4 above)
Comment (xcthulhu) [reply node/src/components/chainspec_loader.rs:251]: As discussed, this functionality will be reintroduced in #1966

File: node/src/components/consensus/error.rs (hunk 20 above)
Comment (goral09) [reply node/src/components/consensus/error.rs:19]: ok

File: node/src/effect/requests.rs (hunk 24 above)
Comment (goral09) [reply node/src/effect/requests.rs:470]: Sure, if you remember it was here.

File: node/CHANGELOG.md (hunk 18 above)
Comment (goral09) [reply node/CHANGELOG.md:35]: Looks like it's still being removed (see `status_feed.rs`).

Comment (xcthulhu) [reply node/CHANGELOG.md:35]: So I'm looking here and it's still in: https://github.com/xcthulhu/casper-node/blob/1ef261249fd8a542cc5607882737665ce055c096/node/src/types/status_feed.rs#L172

Where do you see it missing?

File: node/src/components/linear_chain_sync/operations.rs (hunk 26 above)
Comment (xcthulhu) [reply node/src/components/linear_chain_sync/operations.rs:43]: If we decide to abstract this I can see restoring.  However, since we only use one type for this parameter this is easier to reason about for me...

File: node/src/reactor.rs (hunk 25 above)
Comment (xcthulhu) [reply node/src/reactor.rs:669]: I recently my nightly toolchain and now clippy complains this return isn't necessary.

File: node/CHANGELOG.md (hunk 18 above)
Comment (goral09) [reply node/CHANGELOG.md:35]: Hmm, it was removed when I checked. If it's back, that's fine.

File: utils/nctl/sh/scenarios/configs/gov96.config.toml (hunk 27 above)
Comment (xcthulhu) [reply utils/nctl/sh/scenarios/configs/gov96.config.toml]: This should be fixed in c1d7da3

File: node/src/reactor/joiner/memory_metrics.rs (hunk 15 above)
Comment (xcthulhu) [reply node/src/reactor/joiner/memory_metrics.rs]: As of today, the `MemoryMetrics` structure is defined as follows:

```rust
#[derive(Debug)]
pub(super) struct MemoryMetrics {
    /// Total estimated heap memory usage.
    mem_total: IntGauge,

    /// Estimated heap memory usage of metrics component.
    mem_metrics: IntGauge,
    /// Estimated heap memory usage of network component.
    mem_network: IntGauge,
    /// Estimated heap memory usage of small_network component.
    mem_small_network: IntGauge,
    /// Estimated heap memory usage of address_gossiper component.
    mem_address_gossiper: IntGauge,
    /// Estimated heap memory usage of the configuration for the validator node.
    mem_config: IntGauge,
    /// Estimated heap memory usage for the chainspec loader component.
    mem_chainspec_loader: IntGauge,
    /// Estimated heap memory usage of storage component.
    mem_storage: IntGauge,
    /// Estimated heap memory usage of the contract runtime component.
    mem_contract_runtime: IntGauge,
    /// Estimated heap memory usage of the block fetcher component.
    mem_block_fetcher: IntGauge,
    /// Estimated heap memory usage of deploy fetcher component.
    mem_deploy_fetcher: IntGauge,

    /// Histogram detailing how long it took to estimate memory usage.
    mem_estimator_runtime_s: Histogram,

    /// Instance of registry component to unregister from when being dropped.
    registry: Registry,
}
```

Is this right or do you see some other corrections?

https://github.com/xcthulhu/casper-node/blob/7794ab898ff610b0a762f966060f91d94b21de80/node/src/reactor/joiner/memory_metrics.rs#L8-L40

**Review by Fraser999 — APPROVED** [review]

File: node/src/reactor/joiner/memory_metrics.rs (hunk 15 above)
Comment (xcthulhu) [reply node/src/reactor/joiner/memory_metrics.rs]: It appears the issue was in the `estimate` function elsewhere in this file.

Fixed in 116cbb2099e1de90cabd55f759b3510705ee8b8a

**Review by Fraser999 — APPROVED** [review]

File: .drone.yml [hunk 28]
```diff
@@ -27,12 +27,12 @@ steps:
...
-    commands:
-      - make setup-audit
-      - cargo generate-lockfile
-      - make audit
+#  - name: cargo-audit
+#    image: casperlabs/node-build-u1804
+#    commands:
+#      - make setup-audit
+#      - cargo generate-lockfile
+#      - make audit
```
Comment (dwerner) [inline .drone.yml:35]: Is this commented block intentional?

**Review by fizyk20 — APPROVED** [review]
Looks good as far as I can tell :+1:

**Review by dwerner — APPROVED** [review]
Just one question on some commented drone config.

File: .drone.yml (hunk 28 above)
Comment (xcthulhu) [reply .drone.yml:35]: Yeah, cargo audit is failing on `libp2p` stuff which will get removed once I rebase.
